src/components/Deck.js

```diff
@@ -18,65 +19,99 @@ import { useTransition, animated } from 'react-spring';
+const initialState = {
+  currentSlide: 0,
+  immediate: false,
+  immediateElement: false,
+  currentSlideElement: 0,
+  reverseDirection: false
+};
+
+const defaultSlideEffect = {
+  from: {
+    width: '100%',
+    position: 'absolute',
+    transform: 'translate(100%, 0%)'
+  },
+  enter: {
+    width: '100%',
+    position: 'absolute',
+    transform: 'translate(0, 0%)'
+  },
+  leave: {
+    width: '100%',
+    position: 'absolute',
+    transform: 'translate(-100%, 0%)'
+  },
+  config: { precision: 0 }
+};
 
 const Deck = ({ children, loop, keyboardControls, ...rest }) => {
   // Our default effect for transitioning between slides
-  const defaultSlideEffect = {
-    from: {
-      width: '100%',
-      position: 'absolute',
-      transform: 'translate(100%, 0%)'
-    },
-    enter: {
-      width: '100%',
-      position: 'absolute',
-      transform: 'translate(0, 0%)'
-    },
-    leave: {
-      width: '100%',
-      position: 'absolute',
-      transform: 'translate(-100%, 0%)'
-    },
-    config: { precision: 0 }
-  };
+  const { signal, synchronize } = React.useContext(AnimationMutexContext);
+  const [firstRun, setFirstRun] = React.useState(true);
+
   // Check for slides and then number slides.
   const filteredChildren = Array.isArray(children)
-    ? children
-        // filter if is a Slide
-        .filter(x => isComponentType(x, 'Slide'))
+    ? children.filter(x => isComponentType(x, 'Slide'))
     : console.error('No children passed') || [];
 
-  // return a wrapped slide with the animated.div + style prop curried
-  // and a slideNum prop based on iterator
-
-  const Slides = filteredChildren.map((
-    x,
-    i // eslint-disable-next-line react/display-name
-  ) => ({ style }) => (
-    <animated.div style={{ ...style }}>
-      {{
-        ...x,
-        props: { ...x.props, slideNum: i, keyboardControls }
-      }}
-    </animated.div>
-  ));
+  const slideElementMap = React.useMemo(() => {
+    const map = {};
+    filteredChildren.filter((slide, index) => {
+      map[index] = Array.isArray(slide.props.children)
+        ? slide.props.children.reduce((memo, current) => {
+            if (isComponentType(current, 'SlideElementWrapper')) {
+              memo += 1;
+            }
+            return memo;
+          }, 0)
+        : 0;
+    });
+    return map;
+  }, [filteredChildren]);
 
   // Initialise useDeck hook and get state and dispatch off of it
   const [state, dispatch] = useDeck(
     initialState,
-    Slides.length,
-    loop ? true : false,
-    rest.animationsWhenGoingBack
+    filteredChildren.length,
+    !!loop,
+    rest.animationsWhenGoingBack,
+    slideElementMap
   );
+  const userTransitionEffect =
+    filteredChildren[state.currentSlide].props.transitionEffect || {};
+  const transitionRef = React.useRef();
+
+  React.useEffect(() => {
+    const runSlideTransition = () =>
+      transitionRef.current.start().then(() => signal());
+    if (firstRun) {
```

Comment:
We use a firstRun flag because the first dispatch is based on page load, not key press so we need that first animation call to be synchronized with the mutex.